Feat(design-tokens): Introduce structure of exported tokens to JS (#1101)
lmc-eu/spirit-design-system: packages/web/src/scss/components/Offcanvas/README.md

# Offcanvas

> Offcanvas is a sidebar component that can be toggled via JavaScript to appear from the left, right, or bottom edge of the viewport.

Author: Directions removed.

Buttons or anchors are used as triggers that are attached to specific elements you toggle, and data attributes are used to invoke JavaScript plugin.

> ## JavaScript Plugin

Reviewer: Link to the main readme for how to include JavaScript plugins?

Author: Linked.

For full functionality, you need to provide Spirit JavaScript, which will handle
toggling of the Header component:

```html
<script src="node_modules/@lmc-eu/spirit-web/js/cjs/spirit-web.min.js" async></script>
```

## Usage

### Via data attributes

> Add data-spirit-toggle="offcanvas" and a data-spirit-target or href to the element to automatically assign control of one Offcanvas element.

Author: Example added.

The data-spirit-target attribute accepts a CSS selector to apply the Offcanvas to.

### Via JavaScript

Enable manually with:

```js
const offcanvasElementList = [].slice.call(document.querySelectorAll('.offcanvas'));
const offcanvasList = offcanvasElementList.map(function (offcanvasElement) {
  return new Offcanvas(offcanvasElement);
});
```

### Options

Options can be passed via data attributes or JavaScript.

> For data attributes, append the option name to data-spirit-, as in data-spirit-breakpointDesktop="".

Reviewer: Other multiword breakpoints use the kebab case. Can we please have this as kebab-case too?
eg. `data-spirit-close-on-backdrop-click`

```suggestion
For data attributes, append the option name to data-spirit-, as in data-spirit-breakpoint-desktop="".
```

> For data attributes, append the option name to data-spirit-, as in data-spirit-breakpointDesktop="".

Author: Fixed.

| Name              | Type   | Default | Description                                                       |
| ----------------- | ------ | ------- | ----------------------------------------------------------------- |

> | breakpointDesktop | string | 1280 \* | Desktop breakpoint on which opened Offcanvas automatically closes |

Author: Fixed.

> \*: Default value of the design token `others.breakpoints.desktop`. Can be also changed via `--spirit-breakpoint-desktop` CSS custom property.

Author: Fixed.

### Methods

Activates your content as an offcanvas element. Accepts an optional options object.

You can create an offcanvas instance with the constructor, for example:

```js
const customOffcanvas = document.getElementById('customOffcanvas');
const spiritOffcanvas = new Offcanvas(customOffcanvas);
```

| Method              | Description                                                                                                                                                                                              |
| ------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | toggle              | Toggles an Offcanvas element to shown or hidden. Returns to the caller before the Offcanvas element has actually been shown or hidden (i.e. before the show.offcanvas or hidden.offcanvas event occurs). |

Reviewer: Place add backticks to all items in Method column and in the Event type too.

Author: Added.

# Offcanvas

Offcanvas is a sidebar component that can be toggled via JavaScript to appear from the right edge of the viewport.
Buttons or anchors are used as triggers that are attached to specific elements you toggle, and data attributes are used to invoke JavaScript plugin.

## JavaScript Plugin

For full functionality, you need to [provide Spirit JavaScript][js-plugin-readme], which will handle

> toggling of the Header component:

Reviewer: Header? 

It is first time mentioned :)

Author: Fixed.

```html
<button
  type="button"
  class="Button Button--inverted Button--medium Button--square"
  data-spirit-toggle="offcanvas"
  data-spirit-target="#offcanvas_example"
  aria-controls="offcanvas_example"
  aria-expanded="false"
>
  <svg width="24" height="24" aria-hidden="true">
    <use xlink:href="/icons/svg/sprite.svg#hamburger" />
  </svg>
  <span class="accessibility-hidden">Menu</span>
</button>

<dialog id="offcanvas_example">…</dialog>
```

### Via JavaScript

Enable manually with:

```js
const offcanvasElementList = [].slice.call(document.querySelectorAll('.offcanvas'));
const offcanvasList = offcanvasElementList.map(function (offcanvasElement) {
  return new Offcanvas(offcanvasElement);
});
```

### Options

Options can be passed via data attributes or JavaScript.
For data attributes, append the option name to `data-spirit-`, as in `data-spirit-breakpoint-desktop="1280"`.

| Name                | Type     | Default | Description                                                       |
| ------------------- | -------- | ------- | ----------------------------------------------------------------- |

> | `breakpointDesktop` | `string` | 1280 \* | Desktop breakpoint on which opened Offcanvas automatically closes |

Reviewer: ```suggestion
| `breakpoint-desktop` | `string` | 1280 \* | Desktop breakpoint on which opened Offcanvas automatically closes |
```
?

Author: Fixed.

> \*: Default value of the design token `other.breakpoint.desktop`. Can be also changed via `--spirit-breakpoint-desktop` CSS custom property.

Reviewer: ```suggestion
\*: Default value of the design token `other.breakpoint.desktop`. Can be also changed via `--spirit-breakpoint-desktop` CSS custom property. This CSS property is pre-defined in our foundation CSS.
```

Author: Fixed.

### Events

Spirit's Offcanvas class exposes a few events for hooking into Offcanvas functionality.

| Event type         | Description                                                                                                                  |
| ------------------ | ---------------------------------------------------------------------------------------------------------------------------- |
| `show.offcanvas`   | This event fires immediately when the show instance method is called.                                                        |
| `shown.offcanvas`  | This event is fired when an offcanvas element has been made visible to the user (will wait for CSS transitions to complete). |
| `hide.offcanvas`   | This event is fired immediately when the hide method has been called.                                                        |
| `hidden.offcanvas` | This event is fired when an offcanvas element has been hidden from the user (will wait for CSS transitions to complete)      |

```js
const myOffcanvas = document.getElementById('customOffcanvas');

> customOffcanvas.addEventListener('hidden.offcanvas', function () {

Reviewer: ```suggestion
myOffcanvas.addEventListener('hidden.offcanvas', function () {
```

Author: Fixed.
